Modules/_datetimemodule.c

```diff
@@ -847,6 +847,27 @@ new_date_ex(int year, int month, int day, PyTypeObject *type)
 #define new_date(year, month, day) \
     new_date_ex(year, month, day, &PyDateTime_DateType)
 
+// Forward declaration
+static PyObject * new_datetime_ex(int, int, int, int, int, int, int,
+                                  PyObject*, PyTypeObject*);
+
+/* Create date instance with no range checking, or call subclass constructor */
+static PyObject *
+new_date_subclass_ex(int year, int month, int day, PyObject *cls) {
+    PyObject *result;
+    // We have "fast path" constructors for two subclasses: date and datetime
+    if ( (PyTypeObject *)cls == & PyDateTime_DateType ) {
+        result = new_date_ex(year, month, day, (PyTypeObject *)cls);
+    } else if ((PyTypeObject *)cls == & PyDateTime_DateTimeType ) {
```

Comment:
This is the part I'm not sure about. It feels weird to put special-case logic for a specific subclass in the base class logic, *but* it makes the methods that `datetime` inherits and doesn't re-implement itself (`datetime.today` and `datetime.toordinal`) considerably faster when I do this.

Another option might be to do a token implementation of all the `date` methods specifically calling the `new_datetime_subclass_ex`, or explicitly calling `date.themethod` and then using `datetime.combine` on the result.